sdk/test.go

```diff
@@ -111,30 +101,93 @@ func Test(testReq TestRequest, endpoint string) (*entities.RunResults, error) {
 	return res, nil
 }
 
-func websocketDial(url string, insecure bool, authProtocolType string, authToken string) (*websocket.Conn, *http.Response, error) {
+func websocketDial(urlStr string, insecure bool, authProtocolType string, authToken string) (*websocket.Conn, *http.Response, error) {
+	u, err := transformURL(urlStr)
+	if err != nil {
+		log.Error("error transforming URL: ", err)
+		return nil, nil, err
+	}
+
 	if insecure {
 		websocket.DefaultDialer.TLSClientConfig = &tls.Config{
 			InsecureSkipVerify: true,
 		}
 	}
 
-	str := fmt.Sprintf("* Dialing %s", url)
+	str := fmt.Sprintf("* Dialing %s", u)
 	if insecure {
 		str += " (insecure)"
 	}
 
 	secWebsocketProtocol := http.Header{"Sec-Websocket-Protocol": []string{authProtocolType, authToken}}
 
 	log.Debug(str)
-	c, r, err := websocket.DefaultDialer.Dial(url, secWebsocketProtocol)
+
+	c, r, err := websocket.DefaultDialer.Dial(u, secWebsocketProtocol)
 	if err != nil {
+		log.Error("could not create a new client connection")
 		return nil, r, err
 	}
 
 	log.Debugf("* Websocket connection established")
 	return c, r, nil
 }
 
+func transformURL(urlStr string) (string, error) {
+	u, err := url.Parse(urlStr)
+	if err != nil {
+		return "", err
+	}
+
+	if u.Scheme == "http" {
+		u.Scheme = "ws"
+	} else if u.Scheme == "https" {
+		u.Scheme = "wss"
+	}
+
+	return u.String(), nil
+}
+
+func customError(res *http.Response) error {
+	var e ErrorMsg
+	if err := json.NewDecoder(res.Body).Decode(&e); err != nil {
+		return err
+	}
+	res.Body.Close()
+	return fmt.Errorf("error code: %d, reason: %s", res.StatusCode, e.Error)
+}
+
+func doDial(endpoint string, insecure bool, authProtocolType string, authToken string) (*websocket.Conn, error) {
+	conn, res, err := websocketDial(endpoint, insecure, authProtocolType, authToken)
+	if err == nil {
+		return conn, nil
+	}
+
+	if res == nil {
+		return nil, err
+	}
+
+	if res.StatusCode != 307 {
+		return nil, customError(res)
+	}
+
+	log.Debug("received 307 redirect")
```

Comment:
Are we logging the original url? (And bonus, it might be nice to indent the connection logs)..  For example:

```
Connecting...
* Establishing connection to https://app.capeprivacy.com/
* Received 307 redirect
* Dialing wss://enclave.capeprivacy.com/
```

Reply:
^ Suggesting this for verbose mode